ci/e2e-tests.sh

```diff
@@ -0,0 +1,615 @@
+        )\"
+        if [ -z \"\$ids\" ]; then
+            break
+        else
+            <<< \"\$ids\" xargs az resource delete --ids >/dev/null
+        fi
+
+        sleep 1
+        echo 'Retrying...' >&2
+    done
+
+    if [ -n \"\${vm_os_disk:-}\" ]; then
+        az resource delete --ids \"\$vm_os_disk\"
+    fi
+
+    echo 'Deleted resources' >&2
+" EXIT
+
+
+# Creating an IoT Hub is slow, so parallelize it with prepping the VM.
+
+(
+    echo 'Creating IoT Hub...' >&2
+    iot_hub_resource_id="$(
+        az iot hub create \
+            --resource-group "$AZURE_RESOURCE_GROUP_NAME" \
+            --name "$common_resource_name" \
+            --sku 'S1' --unit 1 --partition-count 2 \
+            --query 'id' --output tsv
+    )"
+    # `az iot hub create` doesn't have `--tags`, so tag it manually.
+    #
+    # Ref: https://github.com/Azure/azure-cli/issues/13497
+    >/dev/null az resource tag \
+        --ids "$iot_hub_resource_id" \
+        --tags "$resource_tag"
+    echo 'Created IoT Hub' >&2
+
+
+    case "$test_name" in
+        'manual-symmetric-key')
+            echo 'Creating IoT device...' >&2
+            iot_device_id="$common_resource_name-01"
+            manual_symmetric_key="$(
+                az iot hub device-identity create \
+                    --hub-name "$common_resource_name" \
+                    --device-id "$iot_device_id" \
+                    --auth-method 'shared_private_key' \
+                    --query 'authentication.symmetricKey.primaryKey' --output tsv
+            )"
+            echo 'Created IoT device' >&2
+
+            echo 'Generating config files...' >&2
+
+            >keyd.toml cat <<-EOF
+[aziot_keys]
+homedir_path = "/var/lib/aziot/keyd"
+
+[preloaded_keys]
+device-id = "file:///var/secrets/aziot/keyd/device-id"
+EOF
+
+            >certd.toml cat <<-EOF
+homedir_path = "/var/lib/aziot/certd"
+
+[cert_issuance]
+
+[preloaded_certs]
+EOF
+
+            >identityd.toml cat <<-EOF
+hostname = "$common_resource_name"
+homedir = "/var/lib/aziot/identityd"
+
+[provisioning]
+dynamic_reprovisioning = true
+source = "manual"
+iothub_hostname = "$common_resource_name.azure-devices.net"
+device_id = "$iot_device_id"
+
+[provisioning.authentication]
+method = "sas"
+device_id_pk = "device-id"
+EOF
+
+            >device-id base64 -d <<< "$manual_symmetric_key"
+
+            echo 'Generated config files' >&2
+            ;;
+
+        *)
+            echo "Unsupported test $1" >&2
+            exit 1
+    esac
+
+    >testmodule.toml cat <<-EOF
+[[principal]]
+uid = 1000
+name = "testmodule"
+idtype = ["module"]
+EOF
+) &
+
+
+echo 'Creating NSG...' >&2
+nsg_id="$(
+    az network nsg create \
+        --resource-group "$AZURE_RESOURCE_GROUP_NAME" \
+        --name "$common_resource_name" \
+        --tags "$resource_tag" \
+        --query 'id' --output tsv
+)"
+echo 'Created NSG' >&2
+
+echo 'Querying public IP...' >&2
+self_ip="$(curl -L 'https://ipinfo.io/ip')"
```

Comment:
I assume this site only provides IPv4 addresses? Since the nsg rule create won't work if this gives IPv6.

Reply:
Just remembered the Azure VMs only have public ipv4 addresses, ignore this.